include/Analysis/RotationAnalysis/RotationAnalysis.h

```diff
@@ -1,215 +1,265 @@
+#include "llvm/include/llvm/Support/Debug.h"               // from @llvm-project
+#include "mlir/include/mlir/Analysis/DataFlowFramework.h"  // from @llvm-project
+#include "mlir/include/mlir/IR/BuiltinTypes.h"             // from @llvm-project
+#include "mlir/include/mlir/IR/Diagnostics.h"              // from @llvm-project
+#include "mlir/include/mlir/IR/Operation.h"                // from @llvm-project
+#include "mlir/include/mlir/IR/OperationSupport.h"         // from @llvm-project
+#include "mlir/include/mlir/IR/Value.h"                    // from @llvm-project
 
 #define DEBUG_TYPE "rotation-analysis"
 
 namespace mlir {
 namespace heir {
 namespace rotation_analysis {
 
-// A wrapper around a mapping from a single tensor SSA value to a set of its
-// access indices.
-class RotationSets {
+// A PartialReduction represents a subset of an arithmetic op tree that reduces
+// values within a tensor to a scalar (present in index zero of the result
+// tensor).
+//
+// It is "partial" in the sense that it may not reduce across all elements of a
+// tensor, and it is used in the analysis to accumulate reduced tensor indices
+// across the IR.
+//
+// It also stores a reference to the SSA value that identifes the "end" of the
+// computation (i.e., the SSA value that contains the result of the reduction).
+class PartialReduction {
  public:
-  enum class Status {
-    // The tensor value has not been set
-    Uninitialized,
-
-    // The rotation set is in a normal state.
-    Normal,
-
-    // The rotation set has a property that makes it invalid for later
-    // optimizations:
-    //
-    //  - It involves operations touch more than one source tensor (not
-    //    including value-semantic outputs)
-    Overdetermined
-
-  };
-
- public:
-  RotationSets() = default;
-  ~RotationSets() = default;
-
-  // Clear the member data, i.e., set the value back to an uninitialized
-  // state.
-  void clear() {
-    accessedIndices.clear();
-    status = Status::Uninitialized;
-  }
-
   bool empty() const { return accessedIndices.empty(); }
 
-  bool isOverdetermined() const { return status == Status::Overdetermined; }
-
-  bool isUninitialized() const { return status == Status::Uninitialized; }
-
   void addRotation(int64_t index) { accessedIndices.insert(index); }
 
-  bool operator==(const RotationSets &rhs) const {
-    return tensor == rhs.tensor && status == rhs.status &&
-           accessedIndices == rhs.accessedIndices;
+  // Returns true if the accessed indices constitute all indices of the reduced
+  // tensor.
+  bool isComplete() const {
+    auto tensorType = tensor.getType().dyn_cast<RankedTensorType>();
+    assert(tensorType &&
+           "Internal state of RotationAnalysis is broken; tensor must have a "
+           "ranked tensor type");
+
+    // std::set is ordered, so min/max is first/last element of the set
+    int64_t minIndex = *accessedIndices.begin();
+    int64_t maxIndex = *accessedIndices.rbegin();
```

Comment:
nit: why use rbegin instead of end if you're just deferencing it?

Reply:
I had to learn this the hard way: `end` points "one past the end" of the container and can't be legally dereferenced.

Reply:
!!! TIL. Thank you!